### PR TITLE
feat: performance improvements for high load

### PR DESCRIPTION
**Pull request by cprice404**
This commit is focused on improving performance for high load,
but also includes some other general cleanups.

Highlights:
* tune grpc memory setting to avoid client-side RESOURCE_EXHAUSTED
* tune grpc subchannel setting to improve performance with multiple
  clients
* SimpleCacheClient now round-robins requests across multiple instances
  of MomentoCache, to spread load across a larger number of grpc channels
* Fixes to logging implementation that were impacting performance and
  stability
* Simplify logging configuration
* Rename files to `kebab-case` to be more idiomatic to typescript
* Fixes all outstanding compiler warnings

For discussion:
* Adds 'request-id' header to requests

### CODE REVIEW

File: src/grpc/headers-interceptor.ts
Comment (cprice404) [inline src/grpc/headers-interceptor.ts:40]: we should talk about this.  I add this because it was useful in helping me debug some things related to retries (otherwise it is very difficult to distinguish requests from one another).  I also assume that it will be generally useful for server-side traces, and possibly for client-side traces/metrics.  But it does add some bytes to the payload, so happy to discuss, and possibly remove / defer to a later time.

File: src/grpc/headers-interceptor.ts
Comment (cprice404) [inline src/grpc/headers-interceptor.ts:45]: this wasn't being set before.  That meant that the agent header was being sent on every request.  I "fixed" it but have mixed feelings about it so would be open to discussing whether this is actually the behavior we want here.

File: src/grpc/retry-interceptor.ts
Comment (cprice404) [inline src/grpc/retry-interceptor.ts:104]: if we decide not to add `request-id` for now then this stuff will need to be removed as well.

File: src/momento-cache.ts
Comment (cprice404) [reply src/momento-cache.ts:58]: did a few more test runs and I think 256 is high enough for now.

File: src/grpc/headers-interceptor.ts
Comment (eaddingtonwhite) [reply src/grpc/headers-interceptor.ts:40]: my 2 cents is its 100% worth the extra bytes having this on every request will make operations much easier prevent us wasting time trouble shooting issues especially now that we are starting to do client side retrys.

Would rather start with this on by default and wait for customer to have issue with bandwidth and then provide an optional way to toggle this header off for very advanced high load users.

I feel similarly about user agent header being sent on every request. Would rather we default to easier operations and reasonable performance and come back and solve for the extreme case when needed and provide ways to toggle off for advanced users.

File: src/grpc/headers-interceptor.ts
Comment (eaddingtonwhite) [reply src/grpc/headers-interceptor.ts:45]: It was performance optimization to cut down on bytes sent. TBH feel its bit of pre-mature optimization see comment on request-id header. That being said I dont care about this one as much vs request-ids so would not be end of world if we keep on only first request like it is now.

File: src/grpc/headers-interceptor.ts
Comment (mjindalmotiv8) [reply src/grpc/headers-interceptor.ts:40]: If we are not tracking this at the service layer I am not sure how we will use this? Is it for client side debugging?

File: src/simple-cache-client.ts
Comment (cprice404) [reply src/simple-cache-client.ts:85]: could do that, let me think about it.  I don't know if we are expecting that customers might ever instantiate this class directly.  I'll keep this in mind when we do the follow-up PR to make this configurable.

File: src/utils/logging.ts
Comment (cprice404) [reply src/utils/logging.ts:41]: interesting.  I can change the default to JSON and just use CONSOLE in the load test then.  Thanks for the heads up.

File: src/grpc/headers-interceptor.ts
Comment (bruuuuuuuce) [reply src/grpc/headers-interceptor.ts:40]: I am 💯 on board for request ids

Comment (danielamiao) [reply src/grpc/headers-interceptor.ts:40]: I'm onboard with adding this but it might be helpful to cross check this with OTel tracing specs (not metrics, don't hate, their tracing protocol is pretty good :P) to ensure we use the standard header tag. There is grpc middleware that automatically links up your traces if you use standard headers (we probably don't have that middleware wrapped around the service yet but we could add it easily later). I can help with this later today if needed.

**Comment by cprice404** [comment]
@eaddingtonwhite do these integration test failures make any sense to you

File: src/grpc/headers-interceptor.ts
Comment (kvcache) [reply src/grpc/headers-interceptor.ts:40]: I'm pretty sure the grpc middleware we have on mr2 already does this @danielamiao 

Agree otlp tracing is not bad at all (otel tracing library though...)

I'm not a fan of doing this for every request. We're a cache, not a database. We should support this but have it default to off - it's an extra 32 bytes per request of overhead paid for by users. Also, it's unnecessary tracking data which we should be cautious with and mindful of.

Comment (bruuuuuuuce) [reply src/grpc/headers-interceptor.ts:40]: > We should support this but have it default to off 

We seem to keep running into these types of issues that require us to go to the customer to tell them to do something so we can debug. I would prefer if we could debug customer requests without having to ask the customer to help us.

File: src/simple-cache-client.ts
Comment (kvcache) [inline src/simple-cache-client.ts:236]:
```suggestion
    const client = this.dataClients[this.nextDataClientIndex];
    this.nextDataClientIndex = (this.nextDataClientIndex + 1) % this.dataClients.length;
    return client;
```
nit: unnecessary condition

**Comment by eaddingtonwhite** [comment]
@cprice404 no that is new one to me have not seen. Id be curious if we ran again if failed in same way consistently.

File: src/grpc/headers-interceptor.ts
Comment (cprice404) [reply src/grpc/headers-interceptor.ts:40]: If we are generally okay with the idea of this existing, but possibly being configurable, we can defer the decision about how to configure it and what the defaults should be until https://github.com/momentohq/dev-eco-issue-tracker/issues/85 , which I intend to pick up very soon.

I don't want to make it configurable in this PR.  For this PR I think the options are either always-on, or remove this code for now.

I can double-check against the otel specs for header name once we have agreed whether or not to even include this in this PR.

File: src/grpc/headers-interceptor.ts
Comment (cprice404) [reply src/grpc/headers-interceptor.ts:45]: need others to weigh in here.  honestly i'd prefer to send this on every request as well.  I'm pretty sure I know what Kenny's opinion is.  Would like to know where @danielamiao , @mjindalmotiv8 , anyone else who is interested in weighing in on this stand.

File: src/grpc/headers-interceptor.ts
Comment (kvcache) [reply src/grpc/headers-interceptor.ts:40]: And you're going to do that with the client's uuid how matt?

It's extra cost footed by the user per request, bloating 100% of requests with tracking information that is extremely debatably useful for 0.000001% of requests.. and again, how are you going to do anything with it without asking the user for something? It's logged client side at debug. If debug logging isn't enabled, then it's completely wasted network transfer - a client side id generator for our service to.. idk what actually

Also, fyi, we do head sampling of request traces. There is a real possibility that uuids a user asks you about will have no trace on an active system and you won't be able to disambiguate "your request never reached us" from "I didn't trace that request"

What is the value offered to users that is worth the cost of 32 bytes additional per request overhead to their cache? Happy to take this to slack

File: src/grpc/headers-interceptor.ts
Comment (mjindalmotiv8) [reply src/grpc/headers-interceptor.ts:45]: I think there is value in this during debugging and having the feature to enable it on per customer selectively during debugging is most likely what we should be doing. Doing on each request is going to add more processing on mr2 and given that we care about microseconds worth of operations, I am worried this will add latency.

**Review by kvcache — DISMISSED** [review]
The request id code as it sits is a bug from my perspective. I understand you're making changes to how the client is configured so whatever change ordering you like seems fine. Ideally we wouldn't merge this as-is to the release branch though if that was the intent.

File: src/grpc/headers-interceptor.ts
Comment (danielamiao) [reply src/grpc/headers-interceptor.ts:40]: @cprice404 let's remove it from this PR to unblock this PR for now. That way we can agree on defaults + use the otlp tracing header in the follow up PR in one go.

File: src/grpc/headers-interceptor.ts
Comment (cprice404) [reply src/grpc/headers-interceptor.ts:45]: configurability is not in scope for this PR, so I will leave this as-is, meaning that it will only be sent once per client.

File: src/grpc/headers-interceptor.ts
Comment (cprice404) [reply src/grpc/headers-interceptor.ts:40]: @danielamiao agreed, I will do that.

Comment (cprice404) [reply src/grpc/headers-interceptor.ts]: removed in https://github.com/momentohq/client-sdk-javascript/pull/84/commits/912471012bc256cf275766de7e83ad7e06efd1e9

File: src/grpc/retry-interceptor.ts
Comment (cprice404) [reply src/grpc/retry-interceptor.ts]: removed in https://github.com/momentohq/client-sdk-javascript/pull/84/commits/912471012bc256cf275766de7e83ad7e06efd1e9

File: src/utils/logging.ts
Comment (cprice404) [reply src/utils/logging.ts]: changed default to JSON in https://github.com/momentohq/client-sdk-javascript/pull/84/commits/912471012bc256cf275766de7e83ad7e06efd1e9

**Comment by cprice404** [comment]
> @cprice404 no that is new one to me have not seen. Id be curious if we ran again if failed in same way consistently.

fixed in https://github.com/momentohq/client-sdk-javascript/pull/84/commits/23034ee0f83cbfa55e794901ef1cd359591c3256 along with some other test cleanups

**Review by kvcache — APPROVED** [review]

**Review by eaddingtonwhite — APPROVED** [review]

**Comment by cprice404** [comment]
@danielamiao anything else you want to see on this one before it goes in?

**Comment by danielamiao** [comment]
@cprice404 nope looks good from my end, thanks!

**Pull request merged**
